src/invoice/verification.rs

```diff
@@ -0,0 +1,389 @@
+        let sig_block = base64::decode(sig.signature.as_bytes())
+            .map_err(|_| SignatureError::CorruptSignature(sig.key.clone()))?;
+
+        let pubkey =
+            PublicKey::from_bytes(&pk).map_err(|_| SignatureError::CorruptKey(sig.key.clone()))?;
+        let ed_sig = EdSignature::new(
+            sig_block
+                .as_slice()
+                .try_into()
+                .map_err(|_| SignatureError::CorruptSignature(sig.key.clone()))?,
+        );
+        pubkey
+            .verify_strict(cleartext, &ed_sig)
+            .map_err(|_| SignatureError::Unverified(sig.key.clone()))
+    }
+    /// Verify that every signature on this invoice is correct.
+    ///
+    /// The verification strategy will determine how this verification is performed.
+    /// Depending on the selected strategy, the `[[signature]]` blocks will be evaluated
+    /// for the following:
+    ///
+    /// - Is the key in the keyring?
+    /// - Can the signature be verified?
+    ///
+    /// Note that the purpose of the keyring is to ensure that we know about the
+    /// entity that claims to have signed the invoice.
+    ///
+    /// If no signatures are on the invoice, this will succeed.
+    ///
+    /// A strategy will determine success or failure based on whether the signature is verified,
+    /// whether the keys are known, whether the requisite number/roles are satisfied, and
+    /// so on.
+    pub fn verify(&self, inv: &Invoice, keyring: &KeyRing) -> Result<(), SignatureError> {
+        let (roles, all_valid, all_verified, all_roles) = match self {
+            VerificationStrategy::GreedyVerification => {
+                (vec![SignatureRole::Creator], true, true, true)
+            }
+            VerificationStrategy::CreativeIntegrity => {
+                (vec![SignatureRole::Creator], false, true, true)
+            }
+            VerificationStrategy::AuthoritativeIntegrity => (
+                vec![SignatureRole::Creator, SignatureRole::Approver],
+                false,
+                false,
+                false,
+            ),
+            VerificationStrategy::ExhaustiveVerification => (
+                vec![
+                    SignatureRole::Creator,
+                    SignatureRole::Approver,
+                    SignatureRole::Host,
+                    SignatureRole::Proxy,
+                ],
+                true,
+                true,
+                false,
+            ),
+            VerificationStrategy::MultipleAttestation(a, b) => (a.clone(), b.clone(), true, true),
+        };
+
+        // Either the Creator or an Approver must be in the keyring
+        match inv.signature.as_ref() {
+            None => Ok(()),
+            Some(signatures) => {
+                let mut known_key = false;
+                let mut filled_roles: Vec<SignatureRole> = vec![];
+                for s in signatures {
+                    debug!(by = %s.by, "Checking signature");
```

Comment:
Super nit: Consider setting up a span for each loop so the `by` value can be set on each logged field

Reply:
Alright, I tried and it didn't seem to work the way I thought. So for now I am gonna skip this.